File: backend/app/models/product.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import Boolean, DateTime, ForeignKey, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)

    farmer_id: Mapped[int] = mapped_column(
        ForeignKey("farmers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    category_id: Mapped[int] = mapped_column(
        ForeignKey("categories.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )

    name: Mapped[str] = mapped_column(
        String(150),
        nullable=False,
        index=True,
    )

    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    price: Mapped[Decimal] = mapped_column(
        Numeric(10, 2),
        nullable=False,
    )

    unit: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    stock_quantity: Mapped[Decimal] = mapped_column(
        Numeric(10, 2),
        nullable=False,
        default=0,
    )

    image: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
    )

    organic: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    is_available: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )

    farmer = relationship(
        "Farmer",
        back_populates="products",
    )

    category = relationship(
        "Category",
        back_populates="products",
    )
    cart_items = relationship(
        "CartItem",
        back_populates="product",
    )

File: frontend/src/types/product.ts
export interface FarmerSummary {
  id: number;
  farm_name: string;
  location: string | null;
  district: string | null;
  state: string | null;
  profile_image: string | null;
  organic_certified: boolean;
  rating: number;
}

export interface Category {
  id: number;
  name: string;
  description?: string | null;
}

export interface Product {
  id: number;
  name: string;
  description: string | null;
  price: string;
  unit: string;
  stock_quantity: string;
  image: string | null;
  organic: boolean;
  is_available: boolean;
  farmer_id: number;
  category_id: number;

  // Available on the product detail API
  farmer?: FarmerSummary;
  category?: Category;
}

export interface ProductListResponse {
  items: Product[];
  total: number;
}
